sql: Apply USING expression to up to one RLS policy

spilchen opened the pull request:
_Stacked PR_

## sql: Add RLS policy information to the optimizer catalog

Row-level security (RLS) policies include expressions that govern table access
during query execution. While policy information was previously added to the
table descriptor, this commit integrates it into the optimizer catalog. The
cat.Table interface has been updated to expose policy details, and a new
cat.Policy interface has been introduced to facilitate interaction with
individual policies.

Epic: CRDB-11724
Informs #136717
Release note: None

## sql: Add RLS statements to the opt test 

Row-level security (RLS) introduced new SQL grammar to manage policies and
enable RLS. This commit implements those statements in the optimizer test
catalog, allowing for optbuilder tests that utilize RLS features.

Newly supported statements:
- **CREATE POLICY:** Creates a new policy for a table.
- **DROP POLICY:** Removes an existing policy from a table.
- **ALTER TABLE .. {ENABLE|DISABLE} ROW LEVEL SECURITY:** Toggles RLS
enforcement for a table.
- **CREATE USER:** Adds support for creating users in the test catalog. RLS
policies are bypassed for admin members.
- **SET ROLE:** Enables changing the current user in optbuilder tests. The test
catalog now tracks the current user and the catalog's users.

Epic: CRDB-11724
Informs #136717
Release note: None

## sql: Apply RLS USING expressions during reads

Row-level security (RLS) allows you to define a USING expression in a policy,
which is applied to all read operations on tables with RLS enabled. A policy
specifies conditions, such as roles and SQL commands, that determine when it
applies. This commit adds support for incorporating a USING expression from at
most one permissive policy into a query.

Currently, the table attribute to enable RLS in the table descriptor is not yet
implemented. As a result, RLS effects are only observable in queries executed
via the optimizer test catalog.

All code paths that drive scans (e.g., SELECT, UPDATE, DELETE) now use
cat.PolicyCommandScope to determine which SQL operations RLS policies apply to.
Certain operations, such as foreign key checks and cascades, are explicitly
exempt from RLS enforcement.

Epic: CRDB-11724
Informs #136717
Release note: None

spilchen committed:
sql: Add RLS policy information to the optimizer catalog

Row-level security (RLS) policies include expressions that govern table access
during query execution. While policy information was previously added to the
table descriptor, this commit integrates it into the optimizer catalog. The
cat.Table interface has been updated to expose policy details, and a new
cat.Policy interface has been introduced to facilitate interaction with
individual policies.

Epic: CRDB-11724
Informs #136717
Release note: None

spilchen committed:
sql: Add RLS statements to the opt test catalog

Row-level security (RLS) introduced new SQL grammar to manage policies and
enable RLS. This commit implements those statements in the optimizer test
catalog, allowing for optbuilder tests that utilize RLS features.

Newly supported statements:
- **CREATE POLICY:** Creates a new policy for a table.
- **DROP POLICY:** Removes an existing policy from a table.
- **ALTER TABLE .. {ENABLE|DISABLE} ROW LEVEL SECURITY:** Toggles RLS
enforcement for a table.
- **CREATE USER:** Adds support for creating users in the test catalog. RLS
policies are bypassed for admin members.
- **SET ROLE:** Enables changing the current user in optbuilder tests. The test
catalog now tracks the current user and the catalog's users.

Epic: CRDB-11724
Informs #136717
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139570)
<!-- Reviewable:end -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great job on this! Looks great! I left a few suggestions.

> Reviewed 22 of 22 files at r1, 14 of 14 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-:-OHhJGdA1dMNIKjbuXBT:b51co75)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h, @rafiss, and @spilchen)

___
*[`pkg/sql/opt/cat/table.go` line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhC6a39Dr9ePEl6B9I:-OHhC6a39Dr9ePEl6B9J:bchjwil) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/cat/table.go#L186)):*
> ```Go
> 
> 	// IsRowLevelSecurityEnabled is true if policies should be applied during the query.
> 	IsRowLevelSecurityEnabled() bool
> ```

Can we simplify the API by having PolicyCount return `0` if RLS is disabled for the table? From the optimizer's perspective, I think this makes sense because it doesn't really care if RLS is enabled or not. It only cares what the policies to apply are, if any.
___
*[`pkg/sql/opt/testutils/testcat/create_policy.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhEuDA0ug0Sl6iV7um:-OHhEuDA0ug0Sl6iV7un:b9yp1gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5834c05580127accd2c8fe1e9d7089ce2db726f2/pkg/sql/opt/testutils/testcat/create_policy.go#L52)):*
> ```Go
> 	}
> 	if n.Exprs.Using != nil {
> 		policy.usingExpr = n.Exprs.Using.String()
> ```

nit: this is going to be re-parsed later on so we should probably use `tree.Serialize(n.Exprs.Using)` like we do here: https://github.com/cockroachdb/cockroach/blob/daf45db54ba1d0e769591dfc68e03f380433296a/pkg/sql/opt/testutils/testcat/create_table.go#L1119
___
*[`pkg/sql/opt/testutils/testcat/create_policy.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhFAsC0DWDGMsbodiA:-OHhFAsD9Ez2WRYVs6tB:bpi03x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5834c05580127accd2c8fe1e9d7089ce2db726f2/pkg/sql/opt/testutils/testcat/create_policy.go#L55)):*
> ```Go
> 	}
> 	if n.Exprs.WithCheck != nil {
> 		policy.withCheckExpr = n.Exprs.WithCheck.String()
> ```

nit: same here, use `tree.Serialize`
___
*[`pkg/sql/opt_catalog.go` line 828 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhB6dj6PX5FXIcem4r:-OHhB6dj6PX5FXIcem4s:b5lffq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt_catalog.go#L828)):*
> ```Go
> 	// Row-level security (RLS) fields
> 	rlsEnabled bool
> 	policies   map[tree.PolicyType][]optPolicy
> ```

nit: you could use two slices instead of a map:

```go
policies struct {
    permissive  []optPolicy
    restrictive []optPolicy
}
```
___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 805 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhCNoh9y2v40DtgMG0:-OHhCNoh9y2v40DtgMG1:bbpnlm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/testutils/testcat/test_catalog.go#L805)):*
> ```Go
> 
> 	rlsEnabled bool
> 	policies   map[tree.PolicyType][]cat.Policy
> ```

nit: you can use two slices here too
___
*[`pkg/sql/opt/optbuilder/fk_cascade.go` line 329 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhJ3jV11PEgHn33ke2:-OHhJ3jV11PEgHn33ke3:b-sm2z54) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b2b7df99a30bf9993a15e3595ff61e211f7121/pkg/sql/opt/optbuilder/fk_cascade.go#L329)):*
> ```Go
> 				b.allocScope(),
> 				true, /* disableNotVisibleIndex */
> 				cat.PolicyScopeExempt,
> ```

nit: Might be nice to add a quick explanation to the comment about `b.buildScan` explaining why we use exempt in all FK and unique check cases.
___
*[`pkg/sql/opt/testutils/testcat/alter_table.go` line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhDqQ-8cBMnwAdUZiK:-OHhDqQ-8cBMnwAdUZiL:b4p9r9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5834c05580127accd2c8fe1e9d7089ce2db726f2/pkg/sql/opt/testutils/testcat/alter_table.go#L102)):*
> ```Go
> 		tt.rlsEnabled = false
> 	default:
> 		return
> ```

nit: Make this branch `panic(errors.AssertionFailed("..."))` to make it obvious that something is wrong if we ever hit this path in a test.
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhGrOj188xynwIm4kz:-OHhGrOj188xynwIm4l-:b-aw55do) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b2b7df99a30bf9993a15e3595ff61e211f7121/pkg/sql/opt/optbuilder/row_level_security.go#L61)):*
> ```Go
> func (b *Builder) buildRowLevelSecurityUsingExpression(
> 	tabMeta *opt.TableMeta, cmdScope cat.PolicyCommandScope,
> ) (expr string) {
> ```

I think it'd be more efficient (and probably simpler) to parse the using expressions in this function and construct the AND-ed and OR-ed filter directly, rather than building a string of AND-ed and OR-ed expressions and parsing that. I think we can just move some of the last lines in the function above into this function and change the return type here to `opt.ScalarExpr`.
___
*[`pkg/sql/opt/cat/policy.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhBtop6raQxmZYlHqq:-OHhBtop6raQxmZYlHqr:b-59fyxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/cat/policy.go#L18)):*
> ```Go
> // is exempt from row-level security. The operations checked must align with the policy
> // commands defined in the CREATE POLICY SQL syntax.
> type PolicyCommandScope int
> ```

nit: use uint8
___
*[`pkg/sql/opt/cat/policy.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhCviG3QXe1cFaLf6Y:-OHhCviG3QXe1cFaLf6Z:b-hnovb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/cat/policy.go#L36)):*
> ```Go
> // Policies use expressions to filter rows during read operations and/or restrict
> // new rows during write operations.
> type Policy interface {
> ```

nit: The two implementations of this interface look identical to me. Can we make this a struct like `testcat.Policy` and use it in both the test catalog and the optCatalog? We do this for `cat.Column`.
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 67 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhIU4TDJbhiAJxixiJ:-OHhIU4TDJbhiAJxixiK:bc4iqya) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b2b7df99a30bf9993a15e3595ff61e211f7121/pkg/sql/opt/optbuilder/testdata/row_level_security#L67)):*
> ```
> ----
> 
> # Verify that a policy for SELECT isn't used for UPDATE 
> ```

nit: remove trailing space and add trailing `.`


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> Reviewed 22 of 22 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139570)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 1874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHi1TNeBf1zq8hnEh4R:-OHi1TNeBf1zq8hnEh4S:b-whmqxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/testutils/testcat/test_catalog.go#L1874)):*
> ```Go
> 
> // Name implements the cat.Policy interface
> func (p *Policy) Name() tree.Name {
> ```

If you do go with Marcus' suggestion and just have a Policy struct instead of an interface, you might also consider getting rid of these accessor methods and just using an exported value.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Apply RLS USING expressions during reads

Row-level security (RLS) allows you to define a USING expression in a policy,
which is applied to all read operations on tables with RLS enabled. A policy
specifies conditions, such as roles and SQL commands, that determine when it
applies. This commit adds support for incorporating a USING expression from at
most one permissive policy into a query.

Currently, the table attribute to enable RLS in the table descriptor is not yet
implemented. As a result, RLS effects are only observable in queries executed
via the optimizer test catalog.

All code paths that drive scans (e.g., SELECT, UPDATE, DELETE) now use
cat.PolicyCommandScope to determine which SQL operations RLS policies apply to.
Certain operations, such as foreign key checks and cascades, are explicitly
exempt from RLS enforcement.

Epic: CRDB-11724
Informs #136717
Release note: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/cat/policy.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhCviG3QXe1cFaLf6Y:-OHiG_CdCpoV7c5Xaaih:bvhvb7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/cat/policy.go#L36)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: The two implementations of this interface look identical to me. Can we make this a struct like `testcat.Policy` and use it in both the test catalog and the optCatalog? We do this for `cat.Column`.
</blockquote></details>

This is a good idea. It cuts down on the amount of duplication.
___
*[`pkg/sql/opt/cat/table.go` line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhC6a39Dr9ePEl6B9I:-OHiTfcKDBalV3R8jbp8:bx22v0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/cat/table.go#L186)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can we simplify the API by having PolicyCount return `0` if RLS is disabled for the table? From the optimizer's perspective, I think this makes sense because it doesn't really care if RLS is enabled or not. It only cares what the policies to apply are, if any.
</blockquote></details>

Yeah, that makes sense.
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHhGrOj188xynwIm4kz:-OHihW2AAo6v7ryVI2wL:b-y0xyb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b2b7df99a30bf9993a15e3595ff61e211f7121/pkg/sql/opt/optbuilder/row_level_security.go#L61)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think it'd be more efficient (and probably simpler) to parse the using expressions in this function and construct the AND-ed and OR-ed filter directly, rather than building a string of AND-ed and OR-ed expressions and parsing that. I think we can just move some of the last lines in the function above into this function and change the return type here to `opt.ScalarExpr`.
</blockquote></details>

Good idea. I agree this will be simpler when we start returning an expression that combines multiple policies.
___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 1874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139570#-OHi1TNeBf1zq8hnEh4R:-OHik_3y2z7eVs7eYKVZ:bdcv6l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca09fe1e99cffa06bef9f7236f2be38abb3e62e4/pkg/sql/opt/testutils/testcat/test_catalog.go#L1874)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

If you do go with Marcus' suggestion and just have a Policy struct instead of an interface, you might also consider getting rid of these accessor methods and just using an exported value.
</blockquote></details>

Good point.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
